cpp/src/gandiva/interval_holder_test.cc

```diff
@@ -0,0 +1,473 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "gandiva/interval_holder.h"
+
+#include <gtest/gtest.h>
+
+#include <memory>
+#include <vector>
+
+#include "gandiva/execution_context.h"
+
+namespace gandiva {
+
+class TestIntervalHolder : public ::testing::Test {
+ protected:
+  ExecutionContext execution_context_;
+};
+
+TEST_F(TestIntervalHolder, TestMatchAllPeriods) {
+  std::shared_ptr<IntervalDaysHolder> interval_days_holder;
+  std::shared_ptr<IntervalYearsHolder> interval_years_holder;
+
+  auto status = IntervalDaysHolder::Make(0, &interval_days_holder);
+  EXPECT_EQ(status.ok(), true) << status.message();
+
+  status = IntervalYearsHolder::Make(0, &interval_years_holder);
+  EXPECT_EQ(status.ok(), true) << status.message();
+
+  auto& cast_interval_day = *interval_days_holder;
+  auto& cast_interval_year = *interval_years_holder;
+
+  // Pass only numbers to cast
+  bool out_valid;
+  std::string data("73834992");
+  int64_t response =
+      cast_interval_day(&execution_context_, data.data(), 8, true, &out_valid);
+  int64_t qty_days_in_response = 0;
+  int64_t qty_millis_in_response = 73834992;
+  EXPECT_TRUE(out_valid);
+  EXPECT_FALSE(execution_context_.has_error());
+  EXPECT_EQ(response, (qty_millis_in_response << 32) | qty_days_in_response);
+
+  int32_t response_interval_yrs =
+      cast_interval_year(&execution_context_, data.data(), 8, true, &out_valid);
+  EXPECT_EQ(response_interval_yrs, 73834992);
+  EXPECT_TRUE(out_valid);
+  EXPECT_FALSE(execution_context_.has_error());
+
+  // Pass only years and days to cast
+  data = "P12Y15D";
```

Comment:
maybe 1 test case with decimal in the final as per rfc?
e.f P12Y15,5D (if that is legal)